File: myFinance_improved_V2/lib/features/inventory_analysis/domain/entities/sales_dashboard.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'sales_dashboard.freezed.dart';

/// 월별 매출 데이터
@freezed
class MonthlyMetrics with _$MonthlyMetrics {
  const factory MonthlyMetrics({
    required num revenue,
    required num margin,
    required num marginRate,
    required int quantity,
  }) = _MonthlyMetrics;
}

/// 성장률 데이터
@freezed
class GrowthMetrics with _$GrowthMetrics {
  const factory GrowthMetrics({
    required num revenuePct,
    required num marginPct,
    required num quantityPct,
  }) = _GrowthMetrics;
}

/// 수익률 대시보드 종합 데이터
/// RPC: get_sales_dashboard 응답
@freezed
class SalesDashboard with _$SalesDashboard {
  const SalesDashboard._();

  const factory SalesDashboard({
    required MonthlyMetrics thisMonth,
    required MonthlyMetrics lastMonth,
    required GrowthMetrics growth,
  }) = _SalesDashboard;

  /// 사업 상태 판정
  /// - good: 매출/마진 모두 양수 성장
  /// - warning: 하나만 양수 성장
  /// - critical: 모두 음수 성장
  String get status {
    final revenueGood = growth.revenuePct > 0;
    final marginGood = growth.marginPct > 0;

    if (revenueGood && marginGood) return 'good';
    if (revenueGood || marginGood) return 'warning';
    return 'critical';
  }

  /// 주의사항 목록 생성
  List<String> get warnings {
    final List<String> result = [];

    if (growth.revenuePct < 0) {
      result.add('매출 ${growth.revenuePct.toStringAsFixed(1)}% 감소');
    }
    if (growth.marginPct < 0) {
      result.add('마진 ${growth.marginPct.toStringAsFixed(1)}% 감소');
    }
    if (growth.quantityPct < -20) {
      result.add('판매량 ${growth.quantityPct.toStringAsFixed(1)}% 급감');
    }
    if (thisMonth.marginRate < 50) {
      result.add('마진율 ${thisMonth.marginRate.toStringAsFixed(1)}% (목표 미달)');
    }

    return result;
  }
}

File: website/src/features/inventory-analysis/domain/entities/salesDashboard.ts
/**
 * Sales Dashboard Domain Entities
 * RPC: get_sales_dashboard
 */

export interface SalesMonthData {
  revenue: number;
  cost: number;
  margin: number;
  marginRate: number;
  quantity: number;
}

export interface SalesGrowth {
  revenuePct: number;
  marginPct: number;
  quantityPct: number;
}

export interface SalesDashboard {
  thisMonth: SalesMonthData;
  lastMonth: SalesMonthData;
  growth: SalesGrowth;
}

/**
 * Map RPC response to domain entity
 */
export function mapSalesDashboardFromRpc(data: Record<string, unknown>): SalesDashboard {
  const thisMonth = data.this_month as Record<string, unknown> | undefined;
  const lastMonth = data.last_month as Record<string, unknown> | undefined;
  const growth = data.growth as Record<string, unknown> | undefined;

  return {
    thisMonth: {
      revenue: Number(thisMonth?.revenue ?? 0),
      cost: Number(thisMonth?.cost ?? 0),
      margin: Number(thisMonth?.margin ?? 0),
      marginRate: Number(thisMonth?.margin_rate ?? 0),
      quantity: Number(thisMonth?.quantity ?? 0),
    },
    lastMonth: {
      revenue: Number(lastMonth?.revenue ?? 0),
      cost: Number(lastMonth?.cost ?? 0),
      margin: Number(lastMonth?.margin ?? 0),
      marginRate: Number(lastMonth?.margin_rate ?? 0),
      quantity: Number(lastMonth?.quantity ?? 0),
    },
    growth: {
      revenuePct: Number(growth?.revenue_pct ?? 0),
      marginPct: Number(growth?.margin_pct ?? 0),
      quantityPct: Number(growth?.quantity_pct ?? 0),
    },
  };
}
